### PR TITLE
Column collapsing menu example#819

### LINKED ISSUE #819: Add an example how to implement a column collapsing menu
As we have the built-in menu in 1.x, it would be great to provide an example how to accomplish a similar thing in 2.0.

### PR DESCRIPTION
**Pull request by niiyeboah**
Fixes #819

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-grid/1560)

### CODE REVIEW

File: demo/grid-columns-demos.html [hunk 1]
```diff
@@ -148,6 +145,22 @@ <h3>Hiding Columns</h3>
...
+            
+            const contextMenu = document.querySelector('vaadin-context-menu');
+            contextMenu.listenOn = document.querySelector('vaadin-button');
+            contextMenu.openOn = 'click';
+            contextMenu.renderer = function(root) {
+              root.innerHTML  = '<div style="display: flex; flex-direction: column;"></div>';
+              columns.forEach((column, i) => {
+                const checkbox = window.document.createElement('vaadin-checkbox');
+                checkbox.textContent = column.getAttribute('name');
+                checkbox.checked = column.hidden;
```
Comment (tomivirkki) [inline demo/grid-columns-demos.html]: Would expect the checked checkboxes should represent the visible columns similarly as here https://demo.vaadin.com/sampler/#ui/grids-and-trees/grid/features

File: demo/grid-columns-demos.html [hunk 2]
```diff
@@ -114,26 +114,23 @@ <h3>Hiding Columns</h3>
...
-          <vaadin-grid-column path="name.last" width="20%" flex-grow="0"></vaadin-grid-column>
-          <vaadin-grid-column path="location.city" width="33%"></vaadin-grid-column>
-          <vaadin-grid-column path="location.state" width="33%"></vaadin-grid-column>
-          <vaadin-grid-column path="location.street" width="33%"></vaadin-grid-column>
+          <vaadin-grid-column width="60px" flex-grow="0" text-align="end" header="#" name="Index"></vaadin-grid-column>
+          <vaadin-grid-column width="60px" flex-grow="0" text-align="center" name="Picture"></vaadin-grid-column>
+          <vaadin-grid-column path="name.first" width="20%" flex-grow="0" name="First Name"></vaadin-grid-column>
+          <vaadin-grid-column path="name.last" width="20%" flex-grow="0" name="Last Name"></vaadin-grid-column>
+          <vaadin-grid-column path="location.city" width="33%" name="City"></vaadin-grid-column>
+          <vaadin-grid-column path="location.state" width="33%" name="State"></vaadin-grid-column>
```
Comment (tomivirkki) [inline demo/grid-columns-demos.html]: One column (maybe "State") could be pre-hidden (with `hidden`) in the demo.

File: demo/grid-columns-demos.html (hunk 1 above)
Comment (niiyeboah) [reply demo/grid-columns-demos.html]: Done

File: demo/grid-columns-demos.html (hunk 2 above)
Comment (niiyeboah) [reply demo/grid-columns-demos.html]: Done

File: demo/grid-columns-demos.html [hunk 3]
```diff
@@ -148,6 +145,22 @@ <h3>Hiding Columns</h3>
               root.firstElementChild.alt = rowData.item.name.first + ' ' + rowData.item.name.last;
               root.firstElementChild.src = rowData.item.picture.thumbnail;
             };
+            
+            const contextMenu = document.querySelector('vaadin-context-menu');
+            contextMenu.listenOn = document.querySelector('vaadin-button');
+            contextMenu.openOn = 'click';
+            contextMenu.renderer = function(root) {
+              root.innerHTML  = '<div style="display: flex; flex-direction: column;"></div>';
+              columns.forEach((column, i) => {
```
Comment (web-padawan) [inline demo/grid-columns-demos.html]: please do not use arrow functions in demos as they are not transpiled on the website and will break in IE11

File: demo/grid-columns-demos.html [hunk 4]
```diff
@@ -114,26 +114,23 @@ <h3>Hiding Columns</h3>
     </p>
     <vaadin-demo-snippet id='grid-columns-demos-hiding-columns'>
       <template preserve-content>
-        <vaadin-checkbox>Hide First Two Columns</vaadin-checkbox>
-
+        <vaadin-context-menu>
+          <vaadin-button aria-label="Hide column menu">
```
Comment (web-padawan) [inline demo/grid-columns-demos.html]: let's add `theme="icon"` to the button so the icon would be aligned nicely.

File: demo/grid-columns-demos.html [hunk 5]
```diff
@@ -148,6 +145,22 @@ <h3>Hiding Columns</h3>
...
+            const contextMenu = document.querySelector('vaadin-context-menu');
+            contextMenu.listenOn = document.querySelector('vaadin-button');
+            contextMenu.openOn = 'click';
+            contextMenu.renderer = function(root) {
+              root.innerHTML  = '<div style="display: flex; flex-direction: column;"></div>';
+              columns.forEach((column, i) => {
+                const checkbox = window.document.createElement('vaadin-checkbox');
+                checkbox.textContent = column.getAttribute('name');
+                checkbox.checked = !column.hidden;
+                checkbox.onchange = () => column.hidden = !checkbox.checked;
```
Comment (web-padawan) [inline demo/grid-columns-demos.html]: let's use `addEventListener` API to listen to `change` and `click` events.

File: demo/grid-columns-demos.html [hunk 6]
```diff
@@ -148,6 +145,22 @@ <h3>Hiding Columns</h3>
               root.firstElementChild.alt = rowData.item.name.first + ' ' + rowData.item.name.last;
               root.firstElementChild.src = rowData.item.picture.thumbnail;
             };
+            
+            const contextMenu = document.querySelector('vaadin-context-menu');
+            contextMenu.listenOn = document.querySelector('vaadin-button');
+            contextMenu.openOn = 'click';
+            contextMenu.renderer = function(root) {
+              root.innerHTML  = '<div style="display: flex; flex-direction: column;"></div>';
```
Comment (web-padawan) [inline demo/grid-columns-demos.html]: we should avoid using `innerHTML` in demos, as it is generally considered unsafe.
Possible alternatives here would be using `display: block` on checkboxes:
```js
checkbox.style.display = 'block';
```

Comment (tomivirkki) [reply demo/grid-columns-demos.html]: In this case it's not that bad as the HTML doesn't include any variables. But, you're right, using `display: block` for the checkboxes might be cleaner. Then the `div` wrapper could be dropped also 👍

File: demo/grid-columns-demos.html (hunk 3 above)
Comment (niiyeboah) [reply demo/grid-columns-demos.html]: I see, I meant to ask about that. Ok I will fix this.

File: demo/grid-columns-demos.html (hunk 6 above)
Comment (niiyeboah) [reply demo/grid-columns-demos.html]: Done

File: demo/grid-columns-demos.html (hunk 5 above)
Comment (niiyeboah) [reply demo/grid-columns-demos.html]: Done

File: demo/grid-columns-demos.html (hunk 4 above)
Comment (niiyeboah) [reply demo/grid-columns-demos.html]: Done

File: demo/grid-columns-demos.html (hunk 3 above)
Comment (niiyeboah) [reply demo/grid-columns-demos.html]: Done

**Pull request merged**